### ISSUE TITLE
Example values for finetuning asr

### ISSUE THREAD

**Issue by YWMditto** [microsoft/SpeechT5#15]
Hi, congratulations on your achievement in this great work! 
It is my first time to use fairseq, so could you please give the exact values or an exmple of the parameters in "ASR finetune" training and inference part， which are these values:

```
DATA_ROOT=
SAVE_DIR=
TRAIN_SET=
VALID_SET=
LABEL_DIR=
BPE_TOKENIZER=
USER_DIR=
PT_CHECKPOINT_PATH=
```

Thanks a lot！
（And the steps that how can get these values would be great!）

**Comment by Ajyy** [comment]
Hi, thanks for your attention!

- DATA_ROOT: path to tsv files, you can use the same tsv files as [wav2vec 2.0](https://github.com/facebookresearch/fairseq/tree/main/examples/wav2vec#prepare-training-data-manifest)

- SAVE_DIR: path for the model checkpoints and tensorboard

- TRAIN_SET: training set, which is the file name of the tsv file without ".tsv"

- VALID_SET: validation set, which is the file name of the tsv file without ".tsv"

- LABEL_DIR: path to the transcriptions (which contains the files with the same file name as tsv file but ended with ".txt")

- BPE_TOKENIZER: bpe tokenizer, which is the [SPM model](https://github.com/microsoft/SpeechT5#language-model-and-vocabulary)

- USER_DIR: which is the path to speecht5 directory, such as SpeechT5/speecht5

- PT_CHECKPOINT_PATH: path of the pre-trained model

Hope this can help you!

**Comment by YWMditto** [comment]
Thanks for your answer, I have looked the code for more details, and understand most of these parameters, 
and first I want to verify that: does the 'BPE_TOKENIZER' means the path to the spm model, such as '/home/spm_char.model'?

And another question is that I see the .tsv file in the mainfest path looks like this:
for example, valid.csv after using the scipt in [https://github.com/facebookresearch/fairseq/tree/main/examples/wav2vec#prepare-training-data-manifest](url)
![image](https://user-images.githubusercontent.com/46778265/195760830-d853642a-deec-4585-ad10-5d53065252d8.png)

So how can I get the corresponding .txt file?

I mean, I have seen the other issue "Text data preparation", but it seems not to be the answer here.

Thanks again for your answer!

**Comment by YWMditto** [comment]
I have seen the processed results using scipt in the word2vec2: https://github.com/facebookresearch/fairseq/blob/main/examples/wav2vec/libri_labels.py

![image](https://user-images.githubusercontent.com/46778265/195769954-ec555406-9b82-492b-be35-790c7dc9cae5.png)

But I see the code of 'SpeechToTextDataset' in SpeechT5, and it seems to need the raw text, so I guess there is a similar script but gather the raw text together.

Should I directly change the 'libri_labels.py'? Or you guys have another used script that could be given to us?

And could I ask what the [ text data preparation](https://github.com/microsoft/SpeechT5/issues/9) is used for? Does it be used for pretraining?

Thanks a lot for taking your time to answer!

**Comment by YWMditto** [comment]
OK, I see the results in the 'valid.wrd' is the raw text....... : ) : ) : )

**Comment by Ajyy** [comment]
For the first question, yes, it's the path to the spm model, such as '/home/spm_char.model'. And I think you got the answer of question 2 : ). Please feel free to ask me if you have further questions.

**Comment by YWMditto** [comment]
Thanks a lot !!!

And the training part of finetuning asr has run , which seems having no problem.
But I encountered a problem in inference part using the asr inference script:

![image](https://user-images.githubusercontent.com/46778265/195809413-40ea1156-2542-4854-9840-7dd486075a59.png)

I have seen the fairseq doc and know that max_tokens means the maximum total num of tokens in one batch.
But I do not understand why and where it happens and how to fix it.

The script's value we use are basiclly as follows:

```
CHECKPOINT_PATH=/home/speecht5_base_asr.pt
DATA_ROOT=/home/dev_clean_mainfest/
SUBSET=valid
BPE_TOKENIZER=/home/spm_char.model
LABEL_DIR=/home/dev_clean_mainfest/
USER_DIR=/home/speecht5
BEAM=4
CTC_WEIGHT=0.5
LM_WEIGHT=0.5
LM_PATH=/home/t5_transformer_lm.pt


fairseq-generate ${DATA_ROOT} \
  --gen-subset ${SUBSET} \
  --bpe-tokenizer ${BPE_TOKENIZER} \
  --user-dir ${USER_DIR} \
  --task speecht5 \
  --t5-task s2t \
  --path ${CHECKPOINT_PATH} \
  --hubert-label-dir ${LABEL_DIR} \
  --ctc-weight ${CTC_WEIGHT} \
  --lm-weight ${LM_WEIGHT} \
  --lm-path ${LM_PATH} \
  --beam ${BEAM} \
  --scoring wer \
  --max-tokens 32000 \
  --max-len-a 0 \
  --max-len-b 620 \
  --sample-rate 16000

```

**Comment by YWMditto** [comment]
ok, it seems a little awkward :), I have seem the max_token is set as 1600000 in training scripts.
Every thing seems different in fairseq, it seems I need more time to adapt.

**Comment by YWMditto** [comment]
It seems so slow using the script above:
![image](https://user-images.githubusercontent.com/46778265/195814550-3e0c46e9-f825-4f4b-a770-229154f2b153.png)

I use a nvidia rtx a5000, and the program  costs about 8-11 GB gpu memory.

Does the inference time and the memory cost normal ?

Thanks for answering so many questions! ₍ᐢ •͈ ༝ •͈ ᐢ₎♡  *⸜( •ᴗ• )⸝*

**Comment by YWMditto** [comment]
And could the finetuned checkpoint of speech asr base model on librispeech 100h  data be used directly?

Cause the result I derived is so odd:

![image](https://user-images.githubusercontent.com/46778265/195824321-79bcfc1a-5ccb-478c-9257-5bed37025506.png)

final output:
![image](https://user-images.githubusercontent.com/46778265/195824659-0e9b439b-24da-461e-8f29-ee0f605ca3d4.png)

I think something has to be wrong. : )

**Comment by Ajyy** [comment]
During inference, if you want to inference with CTC, you need to set the batch size to 1, which means you need to delete the max tokens and set "--batch-size 1".

I suggest that you can firstly verify the performance by setting the weights of CTC and LM to 0. This will be much faster. If the results are OK, then you can try to use CTC and LM during inference.

**Comment by Ajyy** [comment]
You can find the ASR results in [here](https://github.com/microsoft/SpeechT5#automatic-speech-recognition) :)

**Comment by YWMditto** [comment]
okok, let me try try, thanks a lot!

**Comment by YWMditto** [comment]
Hello again, I have see the code of 'SpeechEncoderPrenet' in 'speech_encoder_prenet.py', and it seems when finetuning ASR, the training process still applys the padding mask to the speech input, which is the 'apply_hubert_mask' function.

I am confused because  is not the mask procedure only used in pretraining stage? Why it is used in here, finetuning asr.

Thanks for your reply.

**Comment by Ajyy** [comment]
Hi, we usually apply mask during fine-tuning which is like SpecAugment. You can read the appendix B of [wav2vec2.0 paper](https://arxiv.org/pdf/2006.11477.pdf) for more details

**Comment by YWMditto** [comment]
ok, thanks a lot！

**Comment by YWMditto** [comment]
Hi there, first thanks for answering our so many questions, and here I want to ask a question about the  "Joint pretraining" part in your paper, I was confused for how this part goes.

1. Does this mean that $c_i$ is the quantized vector that is nearest to $u_i$ in the codebook? I mean, not the weighted average of the vectors in the codebook, but only the nearest one.
![image](https://user-images.githubusercontent.com/46778265/196603934-b1019d59-b72c-4e2d-ad9d-00012ec858d4.png)

2. What exactly should be done after the replace operation? Does the `cross-attention` means using these mixed representatoins as encoder output and sending them into decoder to do corresponding reconstruction task?
And also I want to ask what the exactly meaning of $P_k$ is and how it can be computed?
![image](https://user-images.githubusercontent.com/46778265/196604359-d24cc1ba-7ff0-48fa-9031-4bfef2b49b44.png)


Thanks for your reply!!!

**Comment by Ajyy** [comment]
Hi, so sorry for the late reply because I am a little busy for the last two weeks....

1. We actually use the [vector quantizer module](https://github.com/facebookresearch/fairseq/blob/main/fairseq/modules/gumbel_vector_quantizer.py) for our code book. More specifically, a linear layer is used to predict the code, and we replace the encoder output with the corresponding code. You can find more details in our code.
2. Yes, you are right. We use the mixed representation as the input of cross-attention. p_k is from the prediction results as we use a linear layer to predict the code. You can find more details about the loss in the [wav2vec paper](https://arxiv.org/pdf/2006.11477.pdf)

Hope this will help you.

**Comment by YWMditto** [comment]
OK, I have read more papers and find out these could be figured out by reading more, very thanks for your reply for these questions!
I will shut down this issue, thanks a lot!
